[widget audit] toga.MultilineTextInput

freakboy3742 opened the pull request:
## Audit checklist
 - [x] Core tests ported to pytest
 - [x] Core tests at 100% coverage 
 - [x] Core docstrings complete, and in Sphinx format
 - [x] Documentation complete and consistently formatted
 - [x] cocoa backend at 100% coverage 
 - [x] winforms backend at 100% coverage 
 - [x] gtk backend at 100% coverage 
 - [x] iOS backend at 100% coverage 
 - [x] android backend at 100% coverage 
 - [x] Widget support table updated
 - [x] Widget Issue backlog reviewed

freakboy3742 commented on cocoa/src/toga_cocoa/widgets/multilinetextinput.py:
```diff
@@ -14,9 +17,8 @@
     @objc_method
-    def touchBar(self):
-        # Disable the touchbar.
-        return None
```
This was needed in the past to avoid a crash if the app ran on a MacBook with a Touch Bar; however, touchbars are no longer present on modern Macs, and in my testing on an older macBook with a Touch Bar, the bug no longer manifests.

mhsmith reviewed (commented):
The Android implementation is now complete, though there are still some bugs with scrolling and background colors which I'll look at tomorrow.

Meanwhile, here's a partial review:

freakboy3742 commented on examples/multilinetextinput/multilinetextinput/app.py:
```diff
@@ -11,6 +12,20 @@ def enable_toggle_pressed(self, widget, **kwargs):
+            del style.background_color
+        else:
+            style.background_color = YELLOW
+
```
FWIW - I've been using the `colors` app as a testbed for color checks.

mhsmith replied:
OK, in that case I'll remove these buttons.

freakboy3742 commented on core/src/toga/widgets/multilinetextinput.py:
```diff
@@ -1,148 +1,128 @@
         self.on_change = on_change
 
     @property
     def placeholder(self):
-        """The placeholder text.
+        """The placeholder text for the widget.
```
Agreed - I noticed this myself when working on TextInput; I'll wait for you to finish your pass so I don't step on your toes (or, if you feel enthused, feel free to add annotations yourself)

freakboy3742 replied:
Scratch that - I'll do a pass now. Can't step on your toes while you're asleep :-)

freakboy3742 commented on docs/reference/api/widgets/multilinetextinput.rst:
```diff
@@ -1,23 +1,34 @@
 Multi-line text input
 =====================
```
Agreed; updated.

freakboy3742 commented on core/src/toga/widgets/multilinetextinput.py:
```diff
@@ -1,148 +1,128 @@
+        self.refresh()
+
+    @property
+    def enabled(self):
+        """Is the widget currently enabled? i.e., can the user interact with the
+        widget?
+
+        Disabling a MultilineTextInput is equivalent to making the input
+        read-only.
```
Disabling selection as a differentiator between disabled and readonly definitely makes sense (and will have a good analog with TextInput). I'm not sure disabling scrolling makes sense, though. I can't think of any example of a UI where a scroll bar *exists*, but can't be interacted with; from a quick poke around, I can't find any easily exposable API in Cocoa or iOS that would make that simple to implement.

freakboy3742 replied:
I've restored the inherited `enabled` handling for multilinetext, and made the Cocoa implementation a map to `selectable` (since Cocoa is the only platform that doesn't provide a native `enabled` API on multilinetext).

freakboy3742 commented on cocoa/tests_backend/widgets/multilinetextinput.py:
```diff
@@ -1,11 +1,97 @@
+    def has_focus(self):
+        return self.native.window.firstResponder == self.native_text
+
+    @property
+    def visible_height(self):
+        return self.native.contentView.bounds.size.height
+
+    @property
+    def visible_width(self):
+        return self.native.contentView.bounds.size.width
```
I think you're right - visual size should always be the same as the base widget size.

freakboy3742 commented on iOS/tests_backend/widgets/multilinetextinput.py:
```diff
@@ -1,11 +1,76 @@
+    def vertical_scroll_position(self):
+        return self.native.contentOffset.y
+
+    async def wait_for_scroll_completion(self):
+        position = self.vertical_scroll_position
+        current = None
```
Agreed it would be desirable to avoid needing to define an empty method; however, it would require introducing a sleep into platforms that don't need it.

freakboy3742 commented on core/src/toga/widgets/multilinetextinput.py:
```diff
@@ -1,148 +1,128 @@
 
-        Returns:
-            The function ``callable`` that is called on a content change.
+        This is only invoked in response to user-generated changes.
```
Agreed - I got myself turned around here. This has the added benefit of simplifying the GTK implementation as well.

mhsmith commented on android/src/toga_android/widgets/label.py:
```diff
@@ -26,12 +26,6 @@ def set_color(self, value):
-        # On Android, calling setGravity() when the widget has no LayoutParams
-        # results in a NullPointerException.
-        if not self.native.getLayoutParams():
-            return
```
This check was preventing the widget's initial alignment from taking effect. But the testbed and the example app both work fine without it, on both API level 24 (our current minimum) and 33 (the current stable version).

So I guess that either this only affected older versions of Android, or the relative timing of container addition and style application has changed in Toga itself.

freakboy3742 replied:
I'm fairly certain this was a quirk of an early iteration of Toga's layout on Android. There's a bunch of these warts on every widget (usually on the rehint method). I would definitely like to clean these up if we can.

mhsmith commented on android/src/toga_android/widgets/label.py:
```diff
@@ -78,4 +72,4 @@ def rehint(self):
         self.interface.intrinsic.width = at_least(self.native.getMeasuredWidth())
 
     def set_alignment(self, value):
-        self.set_textview_alignment(value, Gravity.CENTER_VERTICAL)
```
Because of the getLayoutParams issue, this line was never previously taking effect during widget creation, and the label remained at the system default alignment of TOP.

After adding a label with an explicit height to the `resize` example, it turns out the vertical alignment of the text is inconsistent between the backends:
* TOP: Android, Cocoa
* CENTER: GTK, iOS, Winforms

In the interest of staying as close as possible to CSS, I suggest we standardize on TOP.

freakboy3742 replied:
Agreed; although we probably want to investigate making this configurable at some point (possibly part of the discussion around #1194). Centered vertical alignment is a very useful option for GUI layout, as it means the centreline of labels is aligned with the centreline of text inputs, widgets, etc.

Actually making the label vertically aligned was trivial on GTK... and more complex than it should be for iOS.

freakboy3742 replied:
Propagating this through to TextInput, it gets a bit more complicated. There's a similar degree of inconsistency about where the text appears in the input... but there's also a near complete lack of control over this feature on macOS, iOS and GTK. macOS uses top alignment by default; iOS and GTK use vertically centered. FWIW, HTML (Safari on macOS) uses vertically centered alignment, but AFAICT there's no way to style this with CSS.

However, IMHO, in the case of TextInput, this is getting to the point where we're overriding platform defaults. In the case of label, it makes sense to have a consistent vertical alignment because we're rendering the location of text in a larger layout, there's no "external" chrome, and setting a height on a label isn't fundamentally unreasonable. But the external chrome of a TextInput will fill the entire allocated space, and should only have a single line of content. Setting a height on a TextInput is generally ill-advised anyway, as a text input should generally be the height of the line of text and no more. I'd argue that if the user is setting an explicit height, it's the operating system's decision where that line of text should appear. 

*If* the platform provides API control for vertical alignment of the text in a TextInput, we can pick a preferred alignment (Center to match HTML?); but if there's no explicit control, then I don't think we should go to extraordinary measures to override the default behavior. 

I think a similar argument could be made for MultilineTextInput - with a more obvious default being TOP. However, I think Android might be the only platform affected by this, and mostly because every widget in Android is apparently a TextLabel :-) Every other platform seems to use TOP for multilinetext, without being configurable.

mhsmith replied:
Agreed: I think MultilineTextInput should definitely be TOP aligned, but setting the height of the single-line TextInput doesn't really make any sense, so it's fine for the alignment of that to be undefined.

mhsmith commented on android/src/toga_android/widgets/webview.py:
```diff
@@ -72,13 +72,6 @@ async def evaluate_javascript(self, javascript):
     def invoke_javascript(self, javascript):
         self.native.evaluateJavascript(str(javascript), ReceiveString())
 
-    def set_alignment(self, value):
-        # Refuse to set alignment unless widget has been added to a container.
-        # This is because this widget's setGravity() requires LayoutParams before it can be called.
-        if not self.native.getLayoutParams():
-            return
-        self.native.setGravity(Gravity.CENTER_VERTICAL | align(value))
```
Because of the getLayoutParams issue, this code was never previously running during widget creation. It turns out WebView doesn't even have a setGravity method.

It doesn't make any sense for WebView to support alignment anyway, so I've just removed the method.

mhsmith commented on android/tests_backend/widgets/label.py:
```diff
@@ -31,3 +31,7 @@ def alignment(self):
         )
         return toga_alignment(self.native.getGravity(), justification_mode)
+
+    @property
+    def vertical_alignment(self):
+        return toga_vertical_alignment(self.native.getGravity())
```
This will have to be implemented by the other backends.

freakboy3742 replied:
I've done this for iOS, Cocoa and GTK; although for iOS and Cocoa, it's a bit of a non-test because there's literally no property to inspect to verify vertical alignment - it's just what the widget does (or, in the case of iOS Label, what the widget has been overridden to do).

freakboy3742 commented:
In the process of debugging a test for NumberInput, I've found a bug with all the text input APIs - we've overloaded the `clear()` API. TextInput and MultilineTextInput both define a `clear()` method that removes all content; but the base Widget also defines `clear()` that removes all children. As TextInput/MultilineTextInput don't have children, there isn't any *functional* issue; but it's obviously a confusing naming overlap.

The `clear()` name was only added recently to Toga (#1893); but the name was inherited from Travertino as a core node operation.

I can see a couple of ways forward:
1. Ignore the problem. It isn't *functionally* ambiguous, as input widgets will always be leaf nodes.
2. Remove `clear()` as an API on text inputs. This would be backwards incompatible, but the functionality is achievable with `self.value = ""`.
3. Rename `clear()` on the base node (and possibly also Travertino) - `clear_children()` perhaps? If we do this, we might need to consider renaming `add`, `insert` and `remove` as well for consistency. If we do this rename, we may also need to re-introduce add/insert/remove/clear APIs for Box as proxies, as this is a commonly used API for constructing layouts.
4. Remove `clear()` as a method on the base node, and only make it available as an API on Box (which actually has children).  Again, for consistency, there might be an argument to be made for removing `add`, `insert` and `remove` as well, as these can only be used on `Box` in practice. This won't hide it from the base node though, unless we also modify Travertino to make these APIs private.

I'm not sure which one I prefer at this point, but I probably fall somewhere between 2 and 3. Thoughts @mhsmith?

mhsmith commented:
I think we should go with 2. Remove `clear()` as an API on text inputs. This method was hardly saving any typing anyway, compared to `value = ""`.

freakboy3742 commented on android/src/toga_android/libs/android/drawable.py:
Should these be here, or in android.graphics?

mhsmith replied:
At some point we should remove the `libs` folder and use Chaquopy's `from ... import ...` syntax, so we don't need to edit 2 files every time we import a new class. To make it easy to do that with a simple search and replace, I've made `drawable` a submodule of `graphics`.

freakboy3742 commented:
> I think we should go with 2. Remove `clear()` as an API on text inputs. This method was hardly saving any typing anyway, compared to `value = ""`.

Done - I've removed it (and added a deprecation note).

freakboy3742 commented:
Success!! I've cleaned up the iOS error - the cause was a weird edge case where the widget already had focus, and was programmatically cleared. The same error was technically present on GTK (and I think Windows) as well, but wasn't visible because of the way the keyboard handler was triggering.

I've also done a small cleanup on the Winforms implementation, removing the need for a proxied copy of the text of the widget; and added a documentation note about the color limitations on Winforms transparent backgrounds.

With that, I'm happy with what is here; if you're happy to, we're good to merge!
